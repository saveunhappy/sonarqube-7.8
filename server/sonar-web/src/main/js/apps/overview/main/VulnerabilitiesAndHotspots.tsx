/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import enhance, { ComposedProps } from './enhance';
import DocTooltip from '../../../components/docs/DocTooltip';
import VulnerabilityIcon from '../../../components/icons-components/VulnerabilityIcon';
import { getMetricName } from '../utils';
import SecurityHotspotIcon from '../../../components/icons-components/SecurityHotspotIcon';

export class VulnerabiltiesAndHotspots extends React.PureComponent<ComposedProps> {
  renderHeader() {
    return this.props.renderHeader('Security');
  }

  renderTimeline(range: string) {
    return this.props.renderTimeline('vulnerabilities', range);
  }

  renderLeak() {
    const { leakPeriod } = this.props;
    if (!leakPeriod) {
      return null;
    }

    return (
      <div className="overview-domain-leak">
        <div className="overview-domain-measures">
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value">
              <span className="offset-left">
                {this.props.renderIssues('new_vulnerabilities', 'VULNERABILITY')}
              </span>
              {this.props.renderRating('new_security_rating')}
            </div>
            <div className="overview-domain-measure-label">
              <VulnerabilityIcon className="little-spacer-right" />
              {getMetricName('new_vulnerabilities')}
            </div>
          </div>
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value overview-domain-measure-value-small">
              <span>{this.props.renderIssues('new_security_hotspots', 'SECURITY_HOTSPOT')}</span>
            </div>
            <div className="overview-domain-measure-label">
              <SecurityHotspotIcon className="little-spacer-right" />
              {getMetricName('new_security_hotspots')}
            </div>
          </div>
        </div>
        {this.renderTimeline('after')}
      </div>
    );
  }

  renderNutshell() {
    return (
      <div className="overview-domain-nutshell">
        <div className="overview-domain-measures">
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value">
              <span className="offset-left">
                {this.props.renderIssues('vulnerabilities', 'VULNERABILITY')}
              </span>
              {this.props.renderRating('security_rating')}
            </div>
            <div className="overview-domain-measure-label display-flex-center display-flex-justify-center">
              <VulnerabilityIcon className="little-spacer-right" />
              {getMetricName('vulnerabilities')}
              <DocTooltip
                className="little-spacer-left"
                doc={import(/* webpackMode: "eager" */ 'Docs/tooltips/metrics/vulnerabilities.md')}
              />
            </div>
            {this.props.renderHistoryLink('vulnerabilities')}
          </div>
          <div className="overview-domain-measure">
            <div className="overview-domain-measure-value overview-domain-measure-value-small">
              {this.props.renderIssues('security_hotspots', 'SECURITY_HOTSPOT')}
            </div>
            <div className="overview-domain-measure-label display-flex-center display-flex-justify-center">
              <SecurityHotspotIcon className="little-spacer-right" />
              {getMetricName('security_hotspots')}
              <DocTooltip
                className="little-spacer-left"
                doc={import(/* webpackMode: "eager" */ 'Docs/tooltips/metrics/security-hotspots.md')}
              />
            </div>
            {this.props.renderHistoryLink('security_hotspots')}
          </div>
        </div>
        {this.renderTimeline('before')}
      </div>
    );
  }

  render() {
    return (
      <div className="overview-card">
        {this.renderHeader()}

        <div className="overview-domain-panel">
          {this.renderNutshell()}
          {this.renderLeak()}
        </div>
      </div>
    );
  }
}

export default enhance(VulnerabiltiesAndHotspots);
