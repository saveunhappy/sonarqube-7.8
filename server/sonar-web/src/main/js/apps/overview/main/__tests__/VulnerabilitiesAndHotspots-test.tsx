/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import VulnerabilitiesAndHotspots from '../VulnerabilitiesAndHotspots';
import { ComposedProps } from '../enhance';
import {
  mockComponent,
  mockMainBranch,
  mockMeasureEnhanced,
  mockMetric
} from '../../../../helpers/testMocks';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();
});

function shallowRender(props: Partial<ComposedProps> = {}) {
  return shallow(
    <VulnerabilitiesAndHotspots
      branchLike={mockMainBranch()}
      component={mockComponent()}
      history={{ vulnerabilities: [] }}
      leakPeriod={{ index: 1 } as T.Period}
      measures={[
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'vulnerabilities',
            key: 'vulnerabilities',
            name: 'Vulnerabilities',
            type: 'INT'
          }),
          value: '0'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'security_hotspots',
            key: 'security_hotspots',
            name: 'Security Hotspots',
            type: 'INT'
          }),
          value: '0'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'security_rating',
            key: 'security_rating',
            name: 'Security',
            type: 'RATING'
          }),
          value: '1.0'
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'new_vulnerabilities',
            key: 'new_vulnerabilities',
            name: 'New Vulnerabilities',
            type: 'INT'
          }),
          periods: [
            {
              bestValue: true,
              index: 1,
              value: '1'
            }
          ]
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'new_security_hotspots',
            key: 'new_security_hotspots',
            name: 'New Security Hotspots',
            type: 'INT'
          }),
          periods: [
            {
              bestValue: true,
              index: 1,
              value: '10'
            }
          ]
        }),
        mockMeasureEnhanced({
          metric: mockMetric({
            id: 'new_security_rating',
            key: 'new_security_rating',
            name: 'New Security Rating',
            type: 'RATING'
          }),
          periods: [
            {
              bestValue: true,
              index: 1,
              value: '5.0'
            }
          ]
        })
      ]}
      {...props}
    />
  ).dive();
}
