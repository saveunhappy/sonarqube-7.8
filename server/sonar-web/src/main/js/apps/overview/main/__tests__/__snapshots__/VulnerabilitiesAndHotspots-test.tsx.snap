// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="overview-card"
>
  <div
    className="overview-card-header"
  >
    <div
      className="overview-title"
    >
      <span>
        metric_domain.Security
      </span>
      <Link
        className="spacer-left small"
        onlyActiveOnIndex={false}
        style={Object {}}
        to={
          Object {
            "pathname": "/component_measures",
            "query": Object {
              "id": "my-project",
              "metric": "Security",
            },
          }
        }
      >
        layout.measures
      </Link>
    </div>
  </div>
  <div
    className="overview-domain-panel"
  >
    <div
      className="overview-domain-nutshell"
    >
      <div
        className="overview-domain-measures"
      >
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value"
          >
            <span
              className="offset-left"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to={
                  Object {
                    "pathname": "/project/issues",
                    "query": Object {
                      "id": "my-project",
                      "resolved": "false",
                      "types": "VULNERABILITY",
                    },
                  }
                }
              >
                0
              </Link>
            </span>
            <Tooltip
              overlay="metric.security_rating.tooltip.A"
            >
              <div
                className="overview-domain-measure-sup"
              >
                <DrilldownLink
                  branchLike={
                    Object {
                      "analysisDate": "2018-01-01",
                      "isMain": true,
                      "name": "master",
                    }
                  }
                  className="link-no-underline"
                  component="my-project"
                  metric="security_rating"
                >
                  <Rating
                    value="1.0"
                  />
                </DrilldownLink>
              </div>
            </Tooltip>
          </div>
          <div
            className="overview-domain-measure-label display-flex-center display-flex-justify-center"
          >
            <VulnerabilityIcon
              className="little-spacer-right"
            />
            overview.metric.vulnerabilities
            <DocTooltip
              className="little-spacer-left"
              doc={Promise {}}
            />
          </div>
          <Link
            className="overview-domain-measure-history-link"
            onlyActiveOnIndex={false}
            style={Object {}}
            to={
              Object {
                "pathname": "/project/activity",
                "query": Object {
                  "custom_metrics": "vulnerabilities",
                  "graph": "custom",
                  "id": "my-project",
                },
              }
            }
          >
            <HistoryIcon />
            <span>
              project_activity.page
            </span>
          </Link>
        </div>
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value overview-domain-measure-value-small"
          >
            <Link
              onlyActiveOnIndex={false}
              style={Object {}}
              to={
                Object {
                  "pathname": "/project/issues",
                  "query": Object {
                    "id": "my-project",
                    "resolved": "false",
                    "types": "SECURITY_HOTSPOT",
                  },
                }
              }
            >
              0
            </Link>
          </div>
          <div
            className="overview-domain-measure-label display-flex-center display-flex-justify-center"
          >
            <SecurityHotspotIcon
              className="little-spacer-right"
            />
            overview.metric.security_hotspots
            <DocTooltip
              className="little-spacer-left"
              doc={Promise {}}
            />
          </div>
          <Link
            className="overview-domain-measure-history-link"
            onlyActiveOnIndex={false}
            style={Object {}}
            to={
              Object {
                "pathname": "/project/activity",
                "query": Object {
                  "custom_metrics": "security_hotspots",
                  "graph": "custom",
                  "id": "my-project",
                },
              }
            }
          >
            <HistoryIcon />
            <span>
              project_activity.page
            </span>
          </Link>
        </div>
      </div>
      <div
        className="overview-domain-timeline"
      >
        <Timeline
          history={Array []}
        />
      </div>
    </div>
    <div
      className="overview-domain-leak"
    >
      <div
        className="overview-domain-measures"
      >
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value"
          >
            <span
              className="offset-left"
            >
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to={
                  Object {
                    "pathname": "/project/issues",
                    "query": Object {
                      "id": "my-project",
                      "resolved": "false",
                      "sinceLeakPeriod": "true",
                      "types": "VULNERABILITY",
                    },
                  }
                }
              >
                1
              </Link>
            </span>
            <Tooltip
              overlay="metric.security_rating.tooltip.E"
            >
              <div
                className="overview-domain-measure-sup"
              >
                <DrilldownLink
                  branchLike={
                    Object {
                      "analysisDate": "2018-01-01",
                      "isMain": true,
                      "name": "master",
                    }
                  }
                  className="link-no-underline"
                  component="my-project"
                  metric="new_security_rating"
                >
                  <Rating
                    value="5.0"
                  />
                </DrilldownLink>
              </div>
            </Tooltip>
          </div>
          <div
            className="overview-domain-measure-label"
          >
            <VulnerabilityIcon
              className="little-spacer-right"
            />
            overview.metric.new_vulnerabilities
          </div>
        </div>
        <div
          className="overview-domain-measure"
        >
          <div
            className="overview-domain-measure-value overview-domain-measure-value-small"
          >
            <span>
              <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to={
                  Object {
                    "pathname": "/project/issues",
                    "query": Object {
                      "id": "my-project",
                      "resolved": "false",
                      "sinceLeakPeriod": "true",
                      "types": "SECURITY_HOTSPOT",
                    },
                  }
                }
              >
                10
              </Link>
            </span>
          </div>
          <div
            className="overview-domain-measure-label"
          >
            <SecurityHotspotIcon
              className="little-spacer-right"
            />
            overview.metric.new_security_hotspots
          </div>
        </div>
      </div>
      <div
        className="overview-domain-timeline"
      >
        <Timeline
          history={Array []}
        />
      </div>
    </div>
  </div>
</div>
`;
